File: bioinformatics/analyze_thermal_consistency.R
# chaipcr/web/public/dynexp/thermal_consistency/analyze.R
# 72C thermal consistency test

library(jsonlite)

# constants
MIN_FLUORESCENCE_VAL <- 800000
MIN_TM_VAL <- 77
MAX_TM_VAL <- 81
MAX_DELTA_TM_VAL <- 2


analyze_thermal_consistency <- function(#floor_temp, # hard-coded inside of the function
    db_usr, db_pwd, db_host, db_port, db_name, 
    exp_id, 
    #stage_id, # hard-coded inside of the function
    calib_info, 
    dcv=FALSE, 
    out_json=TRUE, 
    ... # options to pass onto `mc_tm_pw`
    )
{
    # hard-coded arguments
    stage_id <- 4
    # passed onto `mc_tm_pw`, different than default
    qt_prob <- 0.1
    max_normd_qtv <- 0.9
    channel <- '1'
    
    
    # process the data from only one channel
    
    mc_72c <- process_mc(
        db_usr, db_pwd, db_host, db_port, db_name, 
        exp_id, stage_id, calib_info, 
        dcv=dcv, extra_output=TRUE, 
        qt_prob=qt_prob, max_normd_qtv=max_normd_qtv, # passed onto `mc_tm_pw`, different than default
        ...)
    
    mc_tm <- lapply(mc_72c[['mc_bywell']][[as.character(channel)]], function(well_ele) well_ele[['tm']])
    
    tm_check <- list()
    min_Tm <- 100
    max_Tm <- 0
    
    for (well_name in names(mc_tm)) {
        top1_Tm_Area <- as.list(mc_tm[[well_name]][1,])
        top1_Tm <- top1_Tm_Area[['Tm']]
        if (is.na(top1_Tm)) {
            Tm_bool <- FALSE
        } else {
            if (top1_Tm < min_Tm) min_Tm <- top1_Tm
            if (top1_Tm > max_Tm) max_Tm <- top1_Tm
            Tm_bool <- top1_Tm >= MIN_TM_VAL && top1_Tm <= MAX_TM_VAL }
        top1_Tm_Area[['Tm']] <- list(unbox(top1_Tm), unbox(Tm_bool))
        top1_Tm_Area[['Area']] <- unbox(top1_Tm_Area[['Area']])
        tm_check[[well_name]] <- top1_Tm_Area
        }
    
    delta_Tm <- max_Tm - min_Tm
    
    mc_72c_out<- list(
        'tm_check'=tm_check,
        'delta_Tm'=list(unbox(delta_Tm), unbox(delta_Tm <= MAX_DELTA_TM_VAL))
    )
    
    if (out_json) mc_72c_out <- toJSON(mc_72c_out)
    
    return(mc_72c_out)
    
}

File: bioinformatics/julia/QpcrAnalysis/Essentials/analyze_thermal_consistency.jl
# chaipcr/web/public/dynexp/thermal_consistency/analyze.R
# 72C thermal consistency test

# constants
const MIN_FLUORESCENCE_VAL = 8e5
const MIN_TM_VAL = 77
const MAX_TM_VAL = 81
const MAX_DELTA_TM_VAL = 2


ANALYZE_DICT["thermal_consistency"] = function thermal_consistency(
    # copy over the whole signature of `meltcrv`, comment out `stage_id`
    db_conn::MySQL.MySQLHandle,
    exp_id::Integer, # stage_id::Integer,
    calib_info::Union{Integer,OrderedDict};
    # start: arguments that might be passed by upstream code
    well_nums::AbstractVector=[],
    auto_span_smooth::Bool=false,
    span_smooth_default::Real=0.015,
    span_smooth_factor::Real=7.2,
    # end: arguments that might be passed by upstream code
    dye_in::AbstractString="FAM", dyes_2bfild::AbstractVector=[],
    dcv::Bool=true, # logical, whether to perform multi-channel deconvolution
	max_tmprtr::Real=1000, # maximum temperature to analyze
    # out_format::AbstractString="json", # "full", "pre_json", "json"
    verbose::Bool=false,
    kwdict_mc_tm_pw::OrderedDict=OrderedDict() # keyword arguments passed onto `mc_tm_pw`
    )

    # hard-coded arguments
    stage_id = 4
    # passed onto `mc_tm_pw`, different than default
    qt_prob_flTm = 0.1
    max_normd_qtv = 0.9

    mc_w72c = process_mc(
        db_conn,
        exp_id, stage_id,
        calib_info;
        well_nums=well_nums,
        auto_span_smooth=auto_span_smooth,
        span_smooth_default=span_smooth_default,
        span_smooth_factor=span_smooth_factor,
        dye_in=dye_in, dyes_2bfild=dyes_2bfild,
        dcv=dcv,
        max_tmprtr=max_tmprtr,
        out_format="full",
        verbose=verbose,
        kwdict_mc_tm_pw=OrderedDict(
            :qt_prob_flTm=>qt_prob_flTm,
            :max_normd_qtv=>max_normd_qtv
        )
    )

    # process the data from only one channel
    channel_proc = 1
    channel_proc_i = find(mc_w72c["channels"]) do channel
        channel == channel_proc
    end[1] # do channel

    mc_tm = map(
        mc_bywl -> mc_bywl["Ta_fltd"],
        mc_w72c["mc_bychwl"][:, channel_proc_i]
    )

    tm_check_vec = []
    min_Tm = max_tmprtr + 1
    max_Tm = 0

    for Ta in mc_tm
        if size(Ta)[1] == 0
            tm_check_1w = OrderedDict("Tm" => (NaN, false), "area" => NaN)
        else
            top1_Tm = Ta[1,1]
            if top1_Tm < min_Tm
                min_Tm = top1_Tm
            end
            if top1_Tm > max_Tm
                max_Tm = top1_Tm
            end
            tm_check_1w = OrderedDict(
                "Tm" => (
                    top1_Tm,
                    MIN_TM_VAL <= top1_Tm <= MAX_TM_VAL),
                "area" => Ta[1,2] )
        end # if size
        push!(tm_check_vec, tm_check_1w)
    end # for

    delta_Tm = max_Tm - min_Tm

    mc_w72c_out = OrderedDict(
        "tm_check" => tm_check_vec,
        "delta_Tm" => (round(delta_Tm, JSON_DIGITS), delta_Tm <= MAX_DELTA_TM_VAL)
    )

    return json(mc_w72c_out)

end




# # testing
# module thermal_consistency
# using QpcrAnalysis
# analyze = QpcrAnalysis.analyze_thermal_consistency
# end




#
